Reimplement polling expectations to make them more reliable under heavy system loads

younata opened the pull request:
This trades wall-clock time accuracy for reliability. It's nowhere near as important for a polling expectation to finish within the specified duration as it is to actually poll as expected.

Basically, under heavy system loads, the actual gap between polling attempts increases. Under sufficiently heavy loads, with a sufficiently low timeout, we might never have polled in the first place.

Instead, this calculate how many times we should poll by dividing the timeout by the interval. We then poll exactly that many times, sleeping for however long the polling interval is.

If polling interval is ever 0, then we set it to 1 nanosecond.

WaitUntil is also re-implemented in both async and sync, but that's because the older infrastructure for doing polling was replaced. Starting in Swift 6.3, we will emit a warning when WaitUntil is used in Swift Testing.

This change also removes the ability to detect a blocked runloop, so Nimble will no longer detect for that. In my opinion, this is a worthwhile tradeoff, as blocked runloop was a source of test flakiness.

This isn't changing any public-facing code, aside from adding deprecation warnings. It still counts as a major release.

- [x] Does this have tests?

Fixes #1095, though that's a happy consequence of this.

younata committed:
Reimplement polling expectations to make them more reliable under heavy system loads

This trades wall-clock time accuracy for reliability. It's nowhere near as important for a polling expectation to finish within the specified duration as it is to actually poll as expected.

Basically, under heavy system loads, the actual gap between polling attempts increases. Under sufficiently heavy loads, with a sufficiently low timeout, we might never have polled in the first place.

Instead, this calculate how many times we should poll by dividing the timeout by the interval. We then poll exactly that many times, sleeping for however long the polling interval is.

If polling interval is ever 0, then we set it to 1 nanosecond.

WaitUntil is also re-implemented in both async and sync, but that's because the older infrastructure for doing polling was replaced. Starting in Swift 6.3, we will emit a warning when WaitUntil is used in Swift Testing.

This change also removes the ability to detect a blocked runloop, so Nimble will no longer detect for that. In my opinion, this is a worthwhile tradeoff, as blocked runloop was a source of test flakiness.

younata committed:
Bump xcode version, drop < swift 6.0

younata committed:
Update workflows to specify xcode versions including swift 6

Fix build errors on older xcodes/linux

younata committed:
Fix swiftlint errors

younata committed:
Fix async waitUntil waiting until whenever the closure finishes before returning

younata committed:
Remove unnecessary CoreFoundation import

younata committed:
Update the test watch simulator

younata committed:
Check for more CF imports, and guard it where necessary

younata committed:
Mitigate AsyncAwaitTests.testWaitUntilDoesNotCompleteBeforeRunLoopIsWaiting flakes some more

tahirmt commented:
@younata is there any way to get a release published with this change? This change fixes a ton of flakiness in our test suite of 15k tests.

younata commented:
@tahirmt I still haven't gotten around to addressing @sewerynplazuk's issue. My apologies, I've had a lot going on recently.

younata commented:
https://github.com/Quick/Nimble/pull/1202 addresses the issues that @sewerynplazuk brought up. Once I merge that (tomorrow? It's 11:41 pm for me right now), I'll get version 14 out.
